src/nnbench/types.py

```diff
@@ -93,42 +96,148 @@ def expand(cls, bms: list[dict[str, Any]]) -> BenchmarkRecord:
+        self._path = path
+
+    def load(self):
+        """
+        Returns the path to the artifact on the local file system.
+        """
+        return self._path
+
+
+class S3ArtifactLoader(ArtifactLoader):
+    # TODO: Implement this and other common ArtifactLoders here or in a util
+    pass
+
+
+class Artifact(Generic[T], metaclass=ABCMeta):
     """
     A base artifact class for loading (materializing) artifacts from disk or from remote storage.
 
     This is a helper to convey which kind of type gets loaded for a benchmark in a type-safe way.
     It is most useful when running models on already saved data or models, e.g. when
     comparing a newly trained model against a baseline in storage.
 
-    Subclasses need to implement the `Artifact.materialize()` API, telling nnbench how to
-    load the desired artifact from a path.
+    You need to supply an ArtifactLoader with a load() method to load the Artifact into the
+    local system storage.
+
+    Subclasses need to implement the `Artifact.deserialize()` API, telling nnbench to
+    load the desired artifact from their path.
 
     Parameters
     ----------
-    path: str | os.PathLike[str]
-        Path to the artifact files.
+    loader: ArtifactLoader
+        Loader to get the artifact.
     """
 
-    def __init__(self, path: str | os.PathLike[str]) -> None:
-        # Save the path for later just-in-time materialization.
-        self.path = path
+    def __init__(self, loader: ArtifactLoader) -> None:
+        # Save the path for later just-in-time deserialization.
+        self.path = loader.load()  # fetch the artifact from wherever it resides
         self._value: T | None = None
 
-    @classmethod
-    def materialize(cls) -> "Artifact":
-        """Load the artifact from storage."""
-        raise NotImplementedError
+    @abstractmethod
+    def deserialize(self) -> None:
+        """Deserialize the artifact."""
+
+    def is_deserialized(self) -> bool:
+        """ "Checks if the Artifact is already deserialized."""
+        return self._value is not None
+
+    def __str__(self) -> str:
+        return f"Artifact(path={self.path!r}, is_deserialized={self.is_deserialized()})"
 
+    def __repr__(self) -> str:
+        return f"{self.__class__.__name__}(path={self.path!r}, is_deserialized={self.is_deserialized()})"
+
+    @property
     def value(self) -> T:
+        """
+        Returns the deserialized artifact value.
+
+        Returns
+        -------
+        T
+            The deserialized value of the artifact.
+
+        Raises
+        ------
+        ValueError
+            If the artifact has not been deserialized yet.
+        """
         if self._value is None:
             raise ValueError(
                 f"artifact has not been instantiated yet, "
-                f"perhaps you forgot to call {self.__class__.__name__}.materialize()?"
+                f"perhaps you forgot to call {self.__class__.__name__}.deserialize()?"
             )
```

Comment:
Should this automatically materialize the artifact instead? From a user's point of view, this would make the deserialization completely transparent.

Reply:
I think that makes sense. 
Automatic serialization could also simplify the logic in the `ArtifactCollection` class as its purpose - aside from being an iteration wrapper - is to make sure artifacts are deserialized when we want to use them. 

As you wrote the initial logic, what do you think @nicholasjng ?

Reply:
All good with me.